hanish/darksky.py

```diff
@@ -0,0 +1,192 @@
+# Author:   Benjamin Bengfort <benjamin@bengfort.com>
+# Created:  Mon May 08 12:04:11 2017 -0400
+#
+# Copyright (C) 2016 Bengfort.com
+# For license information, see LICENSE.txt
+#
+# ID: cdarksky.py [] benjamin@bengfort.com $
+
+"""
+Handlers and functionality for making Dark Sky API requests.
+"""
+
+##########################################################################
+## Imports and Module Vars
+##########################################################################
+
+import requests
+
+try:
+    # Python 3
+    from urllib.parse import urljoin
+except ImportError:
+    # Python 2
+    from urlparse import urljoin
+
+
+from datetime import date, datetime
+from expiringdict import ExpiringDict
+from .exceptions import HanishValueError
+from .exceptions import DarkSkyException
+
+
+DARKSKY_API_URL  = "https://api.darksky.net"
+API_CALLS_HEADER = "X-Forecast-API-Calls"
+
+
+##########################################################################
+## DarkSky API
+##########################################################################
+
+class DarkSky(object):
+    """
+    The Dark Sky API object maintains its state of API credentials, number of
+    requests per session, as well as the last time a query was made (cacheing
+    duplicate requests that occur within 5 minutes of each other). Currently,
+    only a portion of the available API methods are implemented, namely the
+    ``forecast`` method, but can be easily extended in the future.
+
+    Parameters
+    ----------
+    apikey: string
+        The API key obtained from the Dark Sky developers service.
+
+    limit: int or None, default = 1000
+        Set a limit to the number of requests made per day. The limit is
+        checked via the X-Forecast-API-Calls header made in API responses,
+        therefore a single request per process will always be made before the
+        limit can be observed.
+
+    cache: time in seconds or None, default = 300
+        Cache requests to a specific zipcode for a specific time limit, to
+        reduce lookups and rate limit queries to the Dark Sky service.
+    """
+
+    def __init__(self, apikey, limit=1000, cache=300):
+        self.apikey      = apikey  # API Key included in each request
+        self.limit       = limit   # Per-day limit (can be None)
+        self.n_api_calls = 0       # Reported number of API queries made
+        self.last_query  = None    # Datetime of the last query made
+        self.cache_timeout = cache # Age in seconds of values in the cache
+
+        # TODO: Create a cache data structure that implements similar
+        # functionality to ``ExpiringDict`` as an internal utility that has no
+        # limit and deals with zip codes and method responses specificially.
+        # Cache is a third party utility, ExpiringDict, which we use here for
+        # speed of development, but should implement ourselves in the future.
+        self.cache = ExpiringDict(max_len=limit, max_age_seconds=cache or 0)
```

Comment:
A good point, my logic uses the timeout as the conditional, and I could certainly test `cache is None` instead. I guess my thought here was that by giving a `max_age_seconds` of 0, developers could access the cache without exceptions (e.g. `NoneType has no attribute get`) , but the cache wouldn't actually cache anything, so if future me or another team member used the cache without checking the timeout, it would still work as expected.